File: lists/linked_list.rb
class LinkedList
  class Node
    attr_accessor :next, :value

    def initialize(value = nil, next_node: nil)
      @value = value
      @next = next_node
    end
  end

  attr_reader :head

  def initialize
    @head = nil
  end

  def find(value)
    return nil unless head
    return head if head.value == value
    find_before(value)&.next
  end

  def find_before(value)
    node = head

    while node&.next
      return node if node.next.value == value
      node = node.next
    end
  end

  def tail
    node = head

    while node
      return node unless node.next
      node = node.next
    end
  end

  def append(value)
    head ? tail.next = Node.new(value) : prepend(value)
  end

  def prepend(value)
    @head = Node.new(value, next_node: head)
  end

  def insert(value, after:)
    node = find(after)
    node.next = Node.new(value, next_node: node.next) if node
  end

  def delete(value)
    return nil unless head
    @head = head.next and return if head.value == value

    node = find_before(value)
    node.next = node.next.next if node
  end
end

File: lists/linked_list.py
class Node:
  def __init__(self, value, next_node):
    self.value = value
    self.next_node = next_node

class LinkedList:
  def __init__(self):
    self.__head = None

  def head(self):
    return self.__head

  def tail(self):
    node = self.__head

    while node:
      if node.next_node:
        node = node.next_node
      else:
        return node

  def append(self, value):
    if self.__head:
      self.tail().next_node = Node(value, None)
    else:
      self.prepend(value)

  def prepend(self, value):
    self.__head = Node(value, self.__head)

  def find(self, value):
    node = self.__head

    while node:
      if node.value == value:
        return node
      else:
        node = node.next_node

  def insert(self, value, after):
    node = self.find(after)

    if node:
      node.next_node = Node(value, node.next_node)

  def delete(self, value):
    if self.__head:
      if self.__head.value == value:
        self.__head = self.__head.next_node
      else:
        node = self.__head

        while node.next_node:
          previous = node
          node = node.next_node

          if node.value == value:
            previous.next_node = node.next_node
            return
